lib/python/ray/worker.py

```diff
@@ -102,6 +102,102 @@ def __del__(self):
+      mapping the names of the reusable variables to the code for reinitializing
+      them. For reusable variables for which reinitializer code is not provided,
+      the reinitializer here essentially wraps the initializer.
+    _used (List[str]): A list of the names of all the reusable variables that
+      have been accessed within the scope of the current task. This is reset to
+      the empty list after each task.
+  """
+  def __init__(self):
+    """Initialize a RayReusables object."""
+    self._names = []
+    self._initializers = {}
+    self._reinitializers = {}
+    self._used = []
+
+  def _reinitialize(self):
+    """Reinitialize the reusable variables that the current task used."""
+    for name in self._used:
+      current_value = self.__getattribute__(name)
+      new_value = self._reinitializers[name](current_value)
+      object.__setattr__(self, name, new_value)
+    self._used = []
+
+  def __getattribute__(self, name):
+    """Get an attribute. This handles reusable variables as a special case.
+
+    When __getattribute__ is called with the name of a reusable variable, that
+    name is added to the list of variables that were used in the current task.
+
+    Args:
+      name (str): The name of the attribute to get.
+    """
+    if name in ["_names", "_initializers", "_reinitializers", "_used", "_reinitialize", "__getattribute__", "__setattr__"]:
+      return object.__getattribute__(self, name)
+    if name in self._names and name not in self._used:
+      self._used.append(name)
+    return object.__getattribute__(self, name)
+
+  def __setattr__(self, name, value):
+    """Set an attribute. This handles reusable variables as a special case.
+
+    This is used to create reusable variables. When it is called, it runs the
+    function for initializing the variable to create the variable. If this is
+    called on the driver, then the functions for initializing and reinitializing
+    the variable are shipped to the workers.
+
+    Args:
+      name (str): The name of the attribute to set. This is either a whitelisted
+        name or it is treated as the name of a reusable variable.
+      value: If name is a whitelisted name, then value can be any value. If name
+        is the name of a reusable variable, then this is either the serialized
+        initializer code or it is a tuple of the serialized initializer and
+        reinitializer code.
+    """
+    if name in ["_names", "_initializers", "_reinitializers", "_used", "_reinitialize", "__getattribute__", "__setattr__"]:
+      object.__setattr__(self, name, value)
+    else:
+      if isinstance(value, tuple):
+        if len(value) != 2:
+          raise Exception("To set a reusable variable, you must pass in an initializer function and a reinitializer function.")
+        initializer, reinitializer = value
+      else:
+        initializer = value
+        reinitializer = lambda val : initializer()
+      if not hasattr(initializer, "__call__"):
```

Comment:
This is the wrong way to test callability; see [this](http://stackoverflow.com/a/16388171/541686).